source/common/network/listener_impl.cc

```diff
@@ -27,23 +31,25 @@ void ListenerImpl::listenCallback(evconnlistener*, evutil_socket_t fd, sockaddr*
                                   int remote_addr_len, void* arg) {
   ListenerImpl* listener = static_cast<ListenerImpl*>(arg);
 
-  Address::InstanceConstSharedPtr final_local_address = listener->socket_.localAddress();
+  // Get the local address from the new socket.  This is different from the listener address, if
+  // the listener is listening on the all hosts address (e.g., 0.0.0.0 for IPv4).
+  Address::InstanceConstSharedPtr final_local_address = listener->getLocalAddress(fd);
```

Comment:
Quick comment: Can we refactor this slightly so that we only do this in the original_dst listener case. I don't think we should pay the syscall penalty for non original dst listeners which is the vastly more common case.